src/zarr/buffer.py

```diff
@@ -0,0 +1,446 @@
+            fill_value
+                If not None, fill the new buffer with a scalar value.
+
+            Return
+            ------
+                A new NDBuffer or subclass instance
+            """
+
+    class NDArrayLike(Protocol):
+        def __call__(self, ndarray_like: NDArrayLike) -> NDBuffer:
+            """Factory function to coerce an array into a NDBuffer (or subclass)
+
+            Callables implementing the `Factor.NDArrayLike` protocol must return
+            an instance of NDBuffer (or subclass) given an ndarray-like object.
+
+            Parameters
+            ----------
+            ndarray_like
+                ndarray-like object
+
+            Return
+            ------
+                A NDBuffer or subclass instance that represents `ndarray_like`
+            """
+
+
+class Buffer:
+    """A flat contiguous memory block
+
+    We use Buffer throughout Zarr to represent a contiguous block of memory.
+
+    A Buffer is backed by a underlying ndarray-like instance that represents
+    the memory. The memory type is unspecified; can be regular host memory,
+    CUDA device memory, or something else. The only requirement is that the
+    ndarray-like instance can be copied/converted to a regular Numpy array
+    (host memory).
+
+    Note
+    ----
+    This buffer is untyped, so all indexing and sizes are in bytes.
+
+    Parameters
+    ----------
+    ndarray_like
+        ndarray-like object that must be 1-dim, contiguous, and byte dtype.
+    """
+
+    def __init__(self, ndarray_like: NDArrayLike):
+        if ndarray_like.ndim != 1:
+            raise ValueError("ndarray_like: only 1-dim allowed")
+        if ndarray_like.dtype != np.dtype("b"):
+            raise ValueError("ndarray_like: only byte dtype allowed")
+        self._data = ndarray_like
+
+    @classmethod
+    def create_zero_length(cls) -> Self:
+        """Create an empty buffer with length zero
+
+        Return
+        ------
+            New empty 0-length buffer
+        """
+        return cls(np.array([], dtype="b"))
+
+    @classmethod
+    def from_ndarray_like(cls, ndarray_like: NDArrayLike) -> Self:
+        """Create a new buffer of a ndarray-like object
+
+        Parameters
+        ----------
+        ndarray_like
+            ndarray-like object that must be 1-dim, contiguous, and byte dtype.
+
+        Return
+        ------
+            New buffer representing `ndarray_like`
+        """
+        return cls(ndarray_like)
+
+    @classmethod
+    def from_bytes(cls, bytes_like: BytesLike) -> Self:
+        """Create a new buffer of a bytes-like object (host memory)
+
+        Parameters
+        ----------
+        bytes_like
+           bytes-like object
+
+        Return
+        ------
+            New buffer representing `bytes_like`
+        """
+        return cls.from_ndarray_like(np.frombuffer(bytes_like, dtype="b"))
+
+    def as_ndarray_like(self) -> NDArrayLike:
```

Comment:
Would it be possible to have an `as_memoryview` method that could be used for passing to Blosc, Gzip etc.? Would that also work without copying?

Reply:
An `as_memoryview` would work like `as_numpy_array`. It might involve copying data depending on what the underlying ndarray-like object the buffer represent. E.g., if `self._data` is a cupy array, we have to copy from device to host memory.

Reply:
I just found the method name `as_ndarray_like` confusing because it actually returns a wrapped 1d byte array

Reply:
Sweet!